### PR TITLE
dynamic: x86_64: Support endbr64 in NOP sled

### PR DESCRIPTION
**Pull request by honggyukim**
Recent gcc might enable -fcf-protection by default, it inserts a special instruction endbr64 at the beginning of each function.

If -fpatchable-function-entry is used together, the NOP instructions are inserted right after endbr64 instruction and it makes our expectation for NOP sled pattern fail as follows.
```
  $ gcc -fpatchable-function-entry=5 -o t-abc s-abc.c
  $ uftrace -P. --debug-domain dynamic:4 t-abc
  dynamic: dynamic patch type: t-abc: 5 (fpatchable)
  dynamic: skip non-applicable functions: a
  dynamic: skip non-applicable functions: b
  dynamic: skip non-applicable functions: c
  dynamic: skip non-applicable functions: main
  # DURATION     TID     FUNCTION
     0.828 us [ 32089] | getpid();
```
To avoid the problem, this patch first checks whether the beginning instruction is 'endbr64' and skips it is matched, then it works fine now.
```
  $ gcc -fpatchable-function-entry=5 -o t-abc s-abc.c
  $ uftrace -P. --debug-domain dynamic:4 t-abc
  dynamic: dynamic patch type: t-abc: 5 (fpatchable)
  dynamic: update 0x55cfdc67616d for 'a' function dynamically to call __fentry__
  dynamic: update 0x55cfdc676184 for 'b' function dynamically to call __fentry__
  dynamic: update 0x55cfdc67619b for 'c' function dynamically to call __fentry__
  dynamic: update 0x55cfdc6761d3 for 'main' function dynamically to call __fentry__
  dynamic: patched all (4) functions in 't-abc'
  # DURATION     TID     FUNCTION
              [ 32366] | main() {
              [ 32366] |   a() {
              [ 32366] |     b() {
              [ 32366] |       c() {
     0.942 us [ 32366] |         getpid();
     2.396 us [ 32366] |       } /* c */
     2.555 us [ 32366] |     } /* b */
     2.701 us [ 32366] |   } /* a */
     3.082 us [ 32366] | } /* main */
```
Fixed: #1613
Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

### CODE REVIEW

File: arch/x86_64/mcount-dynamic.c [hunk 1]
```diff
@@ -151,12 +153,13 @@ static void read_patchable_loc(struct mcount_dynamic_info *mdi, struct uftrace_e
...
 	/* symbol has relative address, fix it to match each other */
-	if (elf->ehdr.e_type == ET_EXEC) {
+	if (elf->ehdr.e_type == ET_EXEC || elf->ehdr.e_type == ET_DYN) {
 		unsigned long *patchable_loc = mdi->patch_target;
 		unsigned i;
 
 		for (i = 0; i < mdi->nr_patch_target; i++) {
-			patchable_loc[i] -= offset;
+			if (elf->ehdr.e_type == ET_EXEC)
+				patchable_loc[i] -= offset;
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c:162]: Hmm.. what's the point of this change?  I think it already supports `ET_DYN`, no?

File: arch/x86_64/mcount-dynamic.c [hunk 2]
```diff
@@ -255,6 +258,10 @@ static int patch_fentry_code(struct mcount_dynamic_info *mdi, struct uftrace_sym
 	unsigned char *insn = (void *)sym->addr + mdi->map->start;
 	unsigned int target_addr;
 
+	/* skip 'endbr64' instruction, which is inserted by (implicit) -fcf-protection option. */
+	if (!memcmp(insn, endbr64, sizeof(endbr64)))
+		insn += sizeof(endbr64);
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c:263]: Yeah, maybe it's easier than changing it to iterate on the `__patchable_function_entries`.

File: arch/x86_64/mcount-dynamic.c (hunk 1 above)
Comment (honggyukim) [reply arch/x86_64/mcount-dynamic.c]: Right, this change is not needed. I've just dropped it.

**Review by namhyung — APPROVED** [review]
LGTM

**Pull request merged**